ethmonitor/ethmonitor.go

```diff
@@ -986,10 +986,15 @@ func (m *Monitor) NumSubscribers() int {
 
 func (m *Monitor) UnsubscribeAll(err error) {
 	m.mu.Lock()
-	defer m.mu.Unlock()
+	var subs []*subscriber
 	for _, sub := range m.subscribers {
+		subs = append(subs, sub)
+	}
+	m.mu.Unlock()
+
+	for _, sub := range subs {
 		sub.err = err
-		sub.Unsubscribe()
+		sub.unsubscribe()
```

Comment:
I'd move locking into `sub.Unsubscribe()`. And this `unsubscribe()` method would be without locking, thus you could lock just once per `UnsubscribeAll()`. Thougths?

Reply:
lock mutex resides in `Monitor` struct. If we want to move lock to `Unsubscribe` then we should have the reference of `Monitor` in `Subscriber` struct as well.